File: mobile-app/lib/models/game.dart
enum ReleaseStatus {
  released,
  earlyAccess,
  openBeta,
  closedBeta,
  alpha,
  comingSoon,
  inDevelopment,
  cancelled;

  String get apiValue {
    switch (this) {
      case ReleaseStatus.released:
        return 'released';
      case ReleaseStatus.earlyAccess:
        return 'early_access';
      case ReleaseStatus.openBeta:
        return 'open_beta';
      case ReleaseStatus.closedBeta:
        return 'closed_beta';
      case ReleaseStatus.alpha:
        return 'alpha';
      case ReleaseStatus.comingSoon:
        return 'coming_soon';
      case ReleaseStatus.inDevelopment:
        return 'in_development';
      case ReleaseStatus.cancelled:
        return 'cancelled';
    }
  }

  String get label {
    switch (this) {
      case ReleaseStatus.released:
        return 'Released';
      case ReleaseStatus.earlyAccess:
        return 'Early Access';
      case ReleaseStatus.openBeta:
        return 'Open Beta';
      case ReleaseStatus.closedBeta:
        return 'Closed Beta';
      case ReleaseStatus.alpha:
        return 'Alpha';
      case ReleaseStatus.comingSoon:
        return 'Coming Soon';
      case ReleaseStatus.inDevelopment:
        return 'In Development';
      case ReleaseStatus.cancelled:
        return 'Cancelled';
    }
  }

  static ReleaseStatus fromString(String? value) {
    switch (value) {
      case 'released':
        return ReleaseStatus.released;
      case 'early_access':
        return ReleaseStatus.earlyAccess;
      case 'open_beta':
        return ReleaseStatus.openBeta;
      case 'closed_beta':
        return ReleaseStatus.closedBeta;
      case 'alpha':
        return ReleaseStatus.alpha;
      case 'coming_soon':
        return ReleaseStatus.comingSoon;
      case 'in_development':
        return ReleaseStatus.inDevelopment;
      case 'cancelled':
        return ReleaseStatus.cancelled;
      default:
        return ReleaseStatus.released;
    }
  }
}

enum AvailabilityStatus {
  available,
  outOfCatalog,
  expiredLicense,
  abandonware,
  publicDomain,
  discontinued,
  rereleased;

  String get apiValue {
    switch (this) {
      case AvailabilityStatus.available:
        return 'available';
      case AvailabilityStatus.outOfCatalog:
        return 'out_of_catalog';
      case AvailabilityStatus.expiredLicense:
        return 'expired_license';
      case AvailabilityStatus.abandonware:
        return 'abandonware';
      case AvailabilityStatus.publicDomain:
        return 'public_domain';
      case AvailabilityStatus.discontinued:
        return 'discontinued';
      case AvailabilityStatus.rereleased:
        return 'rereleased';
    }
  }

  String get label {
    switch (this) {
      case AvailabilityStatus.available:
        return 'Available';
      case AvailabilityStatus.outOfCatalog:
        return 'Out of Catalog';
      case AvailabilityStatus.expiredLicense:
        return 'Expired License';
      case AvailabilityStatus.abandonware:
        return 'Abandonware';
      case AvailabilityStatus.publicDomain:
        return 'Public Domain';
      case AvailabilityStatus.discontinued:
        return 'Discontinued';
      case AvailabilityStatus.rereleased:
        return 'Re-Released';
    }
  }

  static AvailabilityStatus fromString(String? value) {
    switch (value) {
      case 'available':
        return AvailabilityStatus.available;
      case 'out_of_catalog':
        return AvailabilityStatus.outOfCatalog;
      case 'expired_license':
        return AvailabilityStatus.expiredLicense;
      case 'abandonware':
        return AvailabilityStatus.abandonware;
      case 'public_domain':
        return AvailabilityStatus.publicDomain;
      case 'discontinued':
        return AvailabilityStatus.discontinued;
      case 'rereleased':
        return AvailabilityStatus.rereleased;
      default:
        return AvailabilityStatus.available;
    }
  }
}

class Game {
  final int id;
  final String title;
  final String slug;
  final String? description;
  final String? synopsis;
  final int? releaseYear;
  final String? releaseDate;
  final String? coverUrl;
  final String? bannerUrl;
  final String? trailerUrl;
  final ReleaseStatus releaseStatus;
  final AvailabilityStatus availabilityStatus;
  final String? ageRating;
  final double? averageRating;
  final int totalReviews;
  final int? metacriticScore;
  final String? tags;
  final String? developerName;
  final String? publisherName;
  final String? genres;
  final String? platforms;
  final bool isEarlyAccess;
  final bool wasRereleased;
  final DateTime createdAt;
  final DateTime updatedAt;

  Game({
    required this.id,
    required this.title,
    required this.slug,
    this.description,
    this.synopsis,
    this.releaseYear,
    this.releaseDate,
    this.coverUrl,
    this.bannerUrl,
    this.trailerUrl,
    required this.releaseStatus,
    required this.availabilityStatus,
    this.ageRating,
    this.averageRating,
    this.totalReviews = 0,
    this.metacriticScore,
    this.tags,
    this.developerName,
    this.publisherName,
    this.genres,
    this.platforms,
    this.isEarlyAccess = false,
    this.wasRereleased = false,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Game.fromJson(Map<String, dynamic> json) {
    return Game(
      id: json['id'] as int,
      title: json['title'] as String? ?? 'Unknown',
      slug: json['slug'] as String? ?? '',
      description: json['description'] as String?,
      synopsis: json['synopsis'] as String?,
      releaseYear: json['release_year'] as int?,
      releaseDate: json['release_date'] as String?,
      coverUrl: json['cover_url'] as String?,
      bannerUrl: json['banner_url'] as String?,
      trailerUrl: json['trailer_url'] as String?,
      releaseStatus:
          ReleaseStatus.fromString(json['release_status'] as String?),
      availabilityStatus:
          AvailabilityStatus.fromString(json['availability_status'] as String?),
      ageRating: json['age_rating'] as String?,
      averageRating: json['average_rating'] != null
          ? double.tryParse(json['average_rating'].toString())
          : null,
      totalReviews: json['total_reviews'] as int? ?? 0,
      metacriticScore: json['metacritic_score'] as int?,
      tags: json['tags'] as String?,
      developerName: json['developer_name'] as String?,
      publisherName: json['publisher_name'] as String?,
      genres: json['genres'] as String?,
      platforms: json['platforms'] as String?,
      isEarlyAccess: json['is_early_access'] == true,
      wasRereleased: json['was_rereleased'] == true,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'slug': slug,
        'description': description,
        'synopsis': synopsis,
        'release_year': releaseYear,
        'release_date': releaseDate,
        'cover_url': coverUrl,
        'banner_url': bannerUrl,
        'trailer_url': trailerUrl,
        'release_status': releaseStatus.apiValue,
        'availability_status': availabilityStatus.apiValue,
        'age_rating': ageRating,
        'average_rating': averageRating,
        'total_reviews': totalReviews,
        'metacritic_score': metacriticScore,
        'tags': tags,
        'developer_name': developerName,
        'publisher_name': publisherName,
        'genres': genres,
        'platforms': platforms,
        'is_early_access': isEarlyAccess,
        'was_rereleased': wasRereleased,
      };

  /// Returns the cover URL or a placeholder.
  String get displayCoverUrl =>
      coverUrl ?? 'https://via.placeholder.com/300x400?text=${Uri.encodeComponent(title)}';

  /// Returns tags as a list.
  List<String> get tagList =>
      tags?.split(',').map((t) => t.trim()).where((t) => t.isNotEmpty).toList() ?? [];

  /// Returns genres as a list.
  List<String> get genreList =>
      genres?.split(',').map((g) => g.trim()).where((g) => g.isNotEmpty).toList() ?? [];

  /// Returns platforms as a list.
  List<String> get platformList =>
      platforms?.split(',').map((p) => p.trim()).where((p) => p.isNotEmpty).toList() ?? [];
}

File: backend/src/models/Game.ts
import { DataTypes, Model, Optional } from 'sequelize';
import sequelize from '../config/database';

// Game attributes interface
export interface GameAttributes {
  id: number;
  title: string;
  slug: string;
  description?: string;
  synopsis?: string;
  release_year?: number;
  release_date?: Date;
  cover_url?: string;
  banner_url?: string;
  trailer_url?: string;
  developer_id?: number;
  publisher_id?: number;
  release_status: 'released' | 'early_access' | 'open_beta' | 'closed_beta' | 'alpha' | 'coming_soon' | 'in_development' | 'cancelled';
  availability_status: 'available' | 'out_of_catalog' | 'expired_license' | 'abandonware' | 'public_domain' | 'discontinued' | 'rereleased';
  discontinuation_date?: Date;
  official_abandonment_date?: Date;
  rerelease_date?: Date;
  abandonment_reason?: string;
  development_percentage?: number;
  age_rating?: string;
  average_rating?: number;
  total_reviews?: number;
  is_early_access?: boolean;
  was_rereleased?: boolean;
  rawg_id?: number;
  metacritic_score?: number;
  created_at?: Date;
  updated_at?: Date;
}

interface GameCreationAttributes extends Optional<GameAttributes, 'id' | 'description' | 'synopsis' | 'release_year' | 'release_date' | 'cover_url' | 'banner_url' | 'trailer_url' | 'developer_id' | 'publisher_id' | 'discontinuation_date' | 'official_abandonment_date' | 'rerelease_date' | 'abandonment_reason' | 'development_percentage' | 'age_rating' | 'average_rating' | 'total_reviews' | 'is_early_access' | 'was_rereleased' | 'rawg_id' | 'metacritic_score' | 'created_at' | 'updated_at'> { }

class Game extends Model<GameAttributes, GameCreationAttributes> implements GameAttributes {
  public id!: number;
  public title!: string;
  public slug!: string;
  public description?: string;
  public synopsis?: string;
  public release_year?: number;
  public release_date?: Date;
  public cover_url?: string;
  public banner_url?: string;
  public trailer_url?: string;
  public developer_id?: number;
  public publisher_id?: number;
  public release_status!: 'released' | 'early_access' | 'open_beta' | 'closed_beta' | 'alpha' | 'coming_soon' | 'in_development' | 'cancelled';
  public availability_status!: 'available' | 'out_of_catalog' | 'expired_license' | 'abandonware' | 'public_domain' | 'discontinued' | 'rereleased';
  public discontinuation_date?: Date;
  public official_abandonment_date?: Date;
  public rerelease_date?: Date;
  public abandonment_reason?: string;
  public development_percentage?: number;
  public age_rating?: string;
  public average_rating?: number;
  public total_reviews?: number;
  public is_early_access?: boolean;
  public was_rereleased?: boolean;
  public rawg_id?: number;
  public metacritic_score?: number;
  public readonly created_at!: Date;
  public readonly updated_at!: Date;
}

Game.init(
  {
    id: {
      type: DataTypes.INTEGER,
      autoIncrement: true,
      primaryKey: true
    },
    title: {
      type: DataTypes.STRING(255),
      allowNull: false
    },
    slug: {
      type: DataTypes.STRING(255),
      allowNull: false,
      unique: true
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    synopsis: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    release_year: {
      type: DataTypes.INTEGER,
      allowNull: true
    },
    release_date: {
      type: DataTypes.DATEONLY,
      allowNull: true
    },
    cover_url: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    banner_url: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    trailer_url: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    developer_id: {
      type: DataTypes.INTEGER,
      allowNull: true,
      references: {
        model: 'developers',
        key: 'id',
      },
    },
    publisher_id: {
      type: DataTypes.INTEGER,
      allowNull: true,
      references: {
        model: 'publishers',
        key: 'id',
      },
    },
    release_status: {
      type: DataTypes.ENUM('released', 'early_access', 'open_beta', 'closed_beta', 'alpha', 'coming_soon', 'in_development', 'cancelled'),
      defaultValue: 'released',
      allowNull: false
    },
    availability_status: {
      type: DataTypes.ENUM('available', 'out_of_catalog', 'expired_license', 'abandonware', 'public_domain', 'discontinued', 'rereleased'),
      defaultValue: 'available',
      allowNull: false
    },
    discontinuation_date: {
      type: DataTypes.DATEONLY,
      allowNull: true
    },
    official_abandonment_date: {
      type: DataTypes.DATEONLY,
      allowNull: true
    },
    rerelease_date: {
      type: DataTypes.DATEONLY,
      allowNull: true
    },
    abandonment_reason: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    development_percentage: {
      type: DataTypes.INTEGER,
      allowNull: true
    },
    age_rating: {
      type: DataTypes.STRING(20),
      allowNull: true
    },
    average_rating: {
      type: DataTypes.DECIMAL(4, 2),
      allowNull: true
    },
    total_reviews: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
      allowNull: true
    },
    is_early_access: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      allowNull: true
    },
    was_rereleased: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      allowNull: true
    },
    rawg_id: {
      type: DataTypes.INTEGER,
      allowNull: true
    },
    metacritic_score: {
      type: DataTypes.INTEGER,
      allowNull: true
    },
    created_at: {
      type: DataTypes.DATE,
      defaultValue: DataTypes.NOW
    },
    updated_at: {
      type: DataTypes.DATE,
      defaultValue: DataTypes.NOW
    }
  },
  {
    sequelize,
    tableName: 'games',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    indexes: [
      { fields: ['slug'], unique: true },
      { fields: ['title'] },
      { fields: ['developer_id'] },
      { fields: ['publisher_id'] },
      { fields: ['release_status'] },
      { fields: ['availability_status'] },
      { fields: ['release_year'] },
      { fields: ['release_date'] },
      { fields: ['average_rating'] },
      { fields: ['rawg_id'] },
    ],
  }
);

export default Game;
